Add some basic helper functions to Vector2

peppy opened the pull request:
Some minor functional additions to Vector2 bring the base OpenTK fork to parity with our custom fork.

Frassle commented on src/OpenTK/Math/Vector2.cs:
```diff
@@ -68,23 +68,27 @@ public Vector2(float x, float y)
-        public float this[int index] {
-            get{
-                if(index == 0)
+        public float this[int index]
```
Would rather whitespace changes didn't get added with functional changes.

peppy replied:
Yeah... I opted to include this one because visual studio kept automatically applying, but I'll rebase and remove it for now.

Frassle commented on src/OpenTK/Math/Vector2.cs:
```diff
@@ -510,6 +514,28 @@ public static void Clamp(ref Vector2 vec, ref Vector2 min, ref Vector2 max, out
         }
 
         /// <summary>
+        /// Compute the euclidean distance between two vectors.
```
This looks fine, surprised we didn't have something like this already!

varon replied:
Probably subject to the same optimization as listed in the Vector3 PR.

Frassle commented on src/OpenTK/Math/Vector2.cs:
```diff
@@ -695,6 +721,50 @@ public static void Transform(ref Vector2 vec, ref Quaternion quat, out Vector2 r
             result = new Vector2(v.X, v.Y);
         }
 
+        /// <summary>Transform a Position by the given Matrix rotation using right-handed notation.</summary>
```
"using right-handed notation." This doesn't care actually, this is just column vector multiplication you could still pass left-handed coordinates with column vectors.

Nihlus replied:
In this case, I believe it refers to the order of multiplicands, in that the vector is on the right side of the matrix. It should probably be clarified.

varon replied:
@Nihlus - I think that's kinda implicit from the parameter ordering, so I'd rather just leave out mentions of 'handedness' at all.

Frassle commented on src/OpenTK/Math/Vector2.cs:
```diff
@@ -779,6 +849,32 @@ public static void Transform(ref Vector2 vec, ref Quaternion quat, out Vector2 r
+        /// <param name="vec">The vector to transform</param>
+        /// <param name="mat">The desired transformation</param>
+        /// <returns>The transformed vector</returns>
+        public static Vector2 operator *(Matrix3 mat, Vector2 vec)
```
Could be a bit confusing allowing M3*V2. It's an implicit cast of Vector2 to a homogeneous Vector3 and back again. Ok when that's called out explicitly as a transform but it's not really a multiplication.

Nihlus commented:
I would also like unit tests for these new methods to be added. We've got good coverage on Vector2, and I'd like to keep it that way.

Nihlus commented:
@varon I am conceptually for 2, at least in the mathematical portion of the library. However, there are already multiple methods which simplify and abstract away this sort of behaviour in one way or another. I believe we should keep the existing methods, but be stringent about not breaking expected mathematical behaviour (such as matrix size incompatibility, i.e, multiplying a Matrix3 with a Vector2). Instead, we should provide simple ways to do the right thing (such as extracting the correct portion of the matrix and multiplying with that instead).

In this case, adding an opaque operator overload for multiplying a Matrix3 with a Vector2 would go against this way of thinking, but having a way to access the Matrix2 submatrix of the Matrix3 and multiplying with that instead would not.

If I wanted a purely mathematical implementation, I would use System.Numerics.Vectors with perhaps an extension method or two. In graphics, this is a lot of extra work for no real reason. We have a good basis, and as long as we can extend and improve it while not doing things which are apparent contradictions to underlying rules of the math, I believe that is the way we should go.

amulware commented:
I also tend to agree with option 2. Mathematical types like this should not be polluted with not-well-defined operations.

For code that adds functionality that is not purely mathematical, but adds some more geometric (or other) semantics, I would rather use different types, possibly wrapping around the existing ones. Those types could have a 'dumbed down' interface that makes it easier to use for those 99%, but as said, I'm not sure if those would be using OpenTK in the first place.

One could go as far as to have different types for position and direction type vectors, but maybe that's over the top. (Though I like doing this actually, it catches unit-mismatch calculation errors at compile time.)

tzachshabtay commented:
What about having Vector2/3 as the "consistent math" structures, and then adding the "convenience methods" as extension methods (which people can choose not to use)?

Nihlus commented:
If we did that, we'd be better off tacking API onto System.Numerics.Vectors which is another discussion entirely. I'm OK with convenience as long as we're not creating apparent contradictions with established mathematics

tzachshabtay commented:
Hmm, I thought moving to System.Numerics.Vectors is an end goal anyway, for supporting SIMD eventually?

Nihlus commented:
It was discussed in #97 and #401, but I believe issues were encountered on the way. I'd be up for looking into it again once we've split the library, but right now it's just another thing on the backlog.

varon commented:
Concensus is pretty clear then: "Don't break the laws of mathematics but try to make stuff convenient."

@peppy - If you can do the changes that @Frassle and @Nihlus have requested, I think we can do a final look-over then merge.

peppy commented:
Addressed issues mentioned, bar unit tests (I had a quick look but am not too familiar with Xunit so haven't committed yet).

Probably need someone to double-check the new methods – wasn't sure if adding local floats would be better than doing the same float subtraction twice.

Frassle commented:
> wasn't sure if adding local floats would be better than doing the same float subtraction twice.
CSE in the compiler should be able to detect those repeat subtractions fairly easily, it's probably fine.

Looks ok to me.

varon commented:
>Probably need someone to double-check the new methods – wasn't sure if adding local floats would be better than doing the same float subtraction twice.

This kind of thing is going to run through the JIT anyway. Doubt there's any worthwhile performance difference between the two as they don't really have anything to do with memory.

The unit tests are very simple, you should pick up on the syntax in no time, especially if you're just adding tests. Feel free to ping any of the maintainers if you run into trouble.

Nihlus commented on src/OpenTK/Math/Vector2.cs:
```diff
@@ -758,11 +758,31 @@ public static Vector2 Transform(Matrix2 mat, Vector2 pos)
         {
-            result.X = mat.M11 * pos.X + mat.M12 * pos.Y;
-            result.Y = mat.M21 * pos.X + mat.M22 * pos.Y;
+            result.X = mat.Row0.X * pos.X + mat.Row0.Y * pos.Y;
```
Why is the row accessor used here instead of the M** properties?

Nihlus commented on src/OpenTK/Math/Vector2.cs:
```diff
@@ -780,11 +800,31 @@ public static Vector2 Transform(Matrix3 mat, Vector2 pos)
         /// <param name="pos">The position to transform</param>
         /// <param name="result">The transformed vector</param>
-        /// <returns>The transformed position</returns>
         public static void Transform(ref Matrix3 mat, ref Vector2 pos, out Vector2 result)
```
This multiplication is still an undefined operation. For a given matrix with dimensions *m* by *n*, multiplication is only defined for another matrix (which we can consider a vector) of the size *n* by *p*. Therefore, it is undefined for multiplication between a Matrix3 (3x3) and a Vector2 (2x1).

Remove this method.

Nihlus commented on src/OpenTK/Math/Vector2.cs:
```diff
@@ -780,11 +800,31 @@ public static Vector2 Transform(Matrix3 mat, Vector2 pos)
+        /// <param name="pos">The position to transform</param>
+        /// <param name="mat">The desired transformation</param>
+        /// <returns>The transformed position</returns>
+        public static Vector2 Transform(Vector2 pos, Matrix3 mat)
```
This should also be removed.

Nihlus commented on src/OpenTK/Math/Vector2.cs:
```diff
@@ -780,11 +800,31 @@ public static Vector2 Transform(Matrix3 mat, Vector2 pos)
+        /// <param name="pos">The position to transform</param>
+        /// <param name="mat">The desired transformation</param>
+        /// <param name="result">The transformed vector</param>
+        public static void Transform(ref Vector2 pos, ref Matrix3 mat, out Vector2 result)
```
As should this.

peppy commented:
As most of this stuff isn't being accepted (due to being "mathematically incorrect"), we are either going to implement everything as extension methods, or more likely just keep our fork active and used. Unfortunately I do not have time to write the necessary tests for the few methods that remain in this PR, so will close it.

Frassle commented:
I've got some time today to do this quickly. We're still not pulling in the questionable methods but I'll get Distance[Squared] and the Transform by Matrix2 methods in.
